File: src/geek time/c#/week03/105. Construct Binary Tree from Preorder and Inorder Traversal.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace week03 {
    public partial class Solution {
        private IDictionary<int, int> dict_105 = new Dictionary<int, int>();
        public TreeNode BuildTree(int[] preorder, int[] inorder) {
            // 解法：
            // 先序遍历的结果可以看做是： [ [root] [...left children tree...] [...right children tree...] ] 三部分
            // 中序遍历的结果可以看做是： [ [...left children tree...] [root] [...right children tree...] ] 三部分
            
            // 从上面的数组中可以得到如下结论：
            // preorder[p_preorder_start] == root
            // p_preorder_start + 1 == preorder_left_children_tree_start
            // preorder_left_children_tree.Length == p_inorder_root - p_inorder_start  <---  p_inorder_root 需要遍历求出
            // p_preorder_start + preorder_left_children_tree.Length = preorder_left_children_tree_end
            // preorder_left_children_tree_end + 1 == p_preorder_right_children_tree_start
            // p_inorder_root + 1 == inorder_right_children_tree_start

            // 有了上述的各个坐标，就可以通过递归来分别构建不同子树了

            // 首选需要加载中序遍历到Map中加速
            for (var i = 0; i < inorder.Length; i++) {
                dict_105.Add(inorder[i], i);
            }
            return RecursiveBuildingTree(preorder, 0, preorder.Length - 1, inorder, 0, inorder.Length - 1);
        }

        private TreeNode RecursiveBuildingTree(int[] preorder, int p_preorder_start, int p_preorder_end, int[] inorder, int p_inorder_start, int p_inorder_end) {
            // 终止条件
            if (p_preorder_start > p_preorder_end) {
                return null;
            }
            // 构建根节点
            var rootValue = preorder[p_preorder_start];
            var tree = new TreeNode(rootValue);
            var leftChildrenTreeLength = dict_105[rootValue] - p_inorder_start;
            tree.left = RecursiveBuildingTree(preorder, p_preorder_start + 1, p_preorder_start + leftChildrenTreeLength, inorder, p_inorder_start, dict_105[rootValue] - 1);
            tree.right = RecursiveBuildingTree(preorder, p_preorder_start + leftChildrenTreeLength + 1, p_preorder_end, inorder, dict_105[rootValue] + 1, p_inorder_end);
            return tree;
        }
    }
}

File: src/geek time/typescript/week03/105. Construct Binary Tree from Preorder and Inorder Traversal.ts
export class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}

function buildTree(preorder: number[], inorder: number[]): TreeNode | null {
    // 此问题最终其实是转化成为了下标计算的问题
    // 明确可以获取到的坐标为：
    // 1. preorder_root
    // 2. inorder_root
    // 3. left_children_tree_length
    // 4. right_children_tree_start
    const hashMap = new Map<number, number>();
    for (let i = 0; i < inorder.length; i++) {
        hashMap.set(inorder[i], i);
    }
    return myBuildTree(preorder, inorder, 0, preorder.length - 1, 0, inorder.length - 1, hashMap);
}

function myBuildTree(
    preorder: number[],
    inorder: number[],
    p_start: number,
    p_end: number,
    i_start: number,
    i_end: number,
    hashMap: Map<number, number>
): TreeNode | null {
    if (p_start > p_end) return null;
    const rootValue = preorder[p_start];
    const i_root = hashMap.get(rootValue);
    const left_children_tree_length: number = i_root - i_start;
    const tree: TreeNode = new TreeNode(rootValue);
    const left = myBuildTree(preorder, inorder, p_start + 1, p_start + left_children_tree_length, i_start, i_root - 1, hashMap);
    const right = myBuildTree(preorder, inorder, p_start + left_children_tree_length + 1, p_end, i_root + 1, i_end, hashMap);
    tree.left = left;
    tree.right = right;
    return tree;
}
